File: Fields/SalesForce/DropDown/DropDownDesigner.cs
using System;
using System.Linq;
using System.Web.UI;
using Telerik.Sitefinity.Web.UI.ControlDesign;
using System.Collections.Generic;
using Telerik.Sitefinity.Modules.Forms;
using Telerik.Sitefinity.Modules.Forms.Web.UI.Fields;

[assembly: WebResource(SalesForce.Fields.DropDownDesigner.scriptReference, "application/x-javascript")]
namespace SalesForce.Fields
{
    /// <summary>
    /// Represents a designer for the <typeparamref name="SalesForce.Fields.DropDown"/> widget
    /// </summary>
    public class DropDownDesigner : ControlDesignerBase
    {
        #region Properties
        /// <summary>
        /// Obsolete. Use LayoutTemplatePath instead.
        /// </summary>
        protected override string LayoutTemplateName
        {
            get
            {
                return string.Empty;
            }
        }

        /// <summary>
        /// Gets the layout template's relative or virtual path.
        /// </summary>
        public override string LayoutTemplatePath
        {
            get
            {
                if (string.IsNullOrEmpty(base.LayoutTemplatePath))
                    return DropDownDesigner.layoutTemplatePath;
                return base.LayoutTemplatePath;
            }
            set
            {
                base.LayoutTemplatePath = value;
            }
        }

        protected override HtmlTextWriterTag TagKey
        {
            get
            {
                return HtmlTextWriterTag.Div;
            }
        }
        #endregion

        #region Control references
        /// <summary>
        /// Gets the control that is bound to the Title property
        /// </summary>
        protected virtual Control Title
        {
            get
            {
                return this.Container.GetControl<Control>("Title", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the ToolTip property
        /// </summary>
        protected virtual Control Note
        {
            get
            {
                return this.Container.GetControl<Control>("Note", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Description property
        /// </summary>
        protected virtual Control Description
        {
            get
            {
                return this.Container.GetControl<Control>("Description", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the Example property
        /// </summary>
        protected virtual Control Example
        {
            get
            {
                return this.Container.GetControl<Control>("Example", true);
            }
        }

        /// <summary>
        /// Gets the control that is bound to the SalesForceField property
        /// </summary>
        protected virtual Control SalesForceField
        {
            get
            {
                return this.Container.GetControl<Control>("SalesForceField", true);
            }
        }

        protected virtual Control SalesForceFieldType
        {
            get
            {
                return this.Container.GetControl<Control>("SalesForceFieldType", true);
            }
        }
        /// <summary>
        /// Control that represents the meta field name
        /// </summary>
        public MetaFieldNameTextBox MetaFieldNameTextBox
        {
            get { return this.Container.GetControl<MetaFieldNameTextBox>("metaFieldNameTextBox", true); }
        }
        #endregion

        #region Methods
        protected override void InitializeControls(Telerik.Sitefinity.Web.UI.GenericContainer container)
        {
            // Place your initialization logic here
            this.ConfigureMetaFieldName();
        }

        /// <summary>
        /// Configures the name of the meta field.
        /// </summary>
        private void ConfigureMetaFieldName()
        {
            var control = this.PropertyEditor.Control as IFormFieldControl;
            var formControl = (this.PropertyEditor.ControlData as Telerik.Sitefinity.Forms.Model.FormDraftControl);
            var form = formControl.Form;

            // MetaField is not initialized
            if (string.IsNullOrEmpty(control.MetaField.FieldName))
            {
                this.MetaFieldNameTextBox.FieldName = control.GetType().Name + "_" + ((Control)control).ID.ToString();
                this.MetaFieldNameTextBox.ReadOnly = false;
            }
            else
            {
                this.MetaFieldNameTextBox.FieldName = control.MetaField.FieldName;
                if (formControl.Published) // form is already published
                {
                    var fieldName = control.MetaField.FieldName;
                    var fieldExists = form.GetMetaFields().Any(m => m.FieldName == fieldName);
                    if (fieldExists)
                    {
                        this.MetaFieldNameTextBox.ReadOnly = true;
                    }
                }
            }
        }
        #endregion

        #region IScriptControl implementation
        /// <summary>
        /// Gets a collection of script descriptors that represent ECMAScript (JavaScript) client components.
        /// </summary>
        public override System.Collections.Generic.IEnumerable<System.Web.UI.ScriptDescriptor> GetScriptDescriptors()
        {
            var scriptDescriptors = new List<ScriptDescriptor>(base.GetScriptDescriptors());
            var descriptor = (ScriptControlDescriptor)scriptDescriptors.Last();

            descriptor.AddElementProperty("title", this.Title.ClientID);
            descriptor.AddElementProperty("note", this.Note.ClientID);
            descriptor.AddElementProperty("description", this.Description.ClientID);
            descriptor.AddElementProperty("example", this.Example.ClientID);
            descriptor.AddElementProperty("salesForceField", this.SalesForceField.ClientID);
            descriptor.AddElementProperty("salesForceFieldType", this.SalesForceFieldType.ClientID);
            descriptor.AddComponentProperty("metaFieldNameTextBox", this.MetaFieldNameTextBox.ClientID);
            return scriptDescriptors.ToArray();
        }

        /// <summary>
        /// Gets a collection of ScriptReference objects that define script resources that the control requires.
        /// </summary>
        public override System.Collections.Generic.IEnumerable<System.Web.UI.ScriptReference> GetScriptReferences()
        {
            var scripts = new List<ScriptReference>(base.GetScriptReferences());
            scripts.Add(new ScriptReference(DropDownDesigner.scriptReference, typeof(DropDownDesigner).Assembly.FullName));
            return scripts;
        }
        #endregion

        #region Private members & constants
        public static readonly string layoutTemplatePath = "~/SalesForce.Fields/Templates/DropDownDesigner.ascx";
        public const string scriptReference = "SalesForce.Fields.Fields.SalesForce.DropDown.DropDownDesigner.js";
        #endregion
    }
}
 

File: Fields/SalesForce/DropDown/DropDownDesigner.js
Type.registerNamespace("SalesForce.Fields");

SalesForce.Fields.DropDownDesigner = function (element) {
    /* Initialize Title fields */
    this._title = null;
    
    /* Initialize ToolTip fields */
    this._note = null;
    
    /* Initialize Description fields */
    this._description = null;
    
    /* Initialize Example fields */
    this._example = null;
    
    /* Initialize SalesForceField fields */
    this._salesForceField = null;

    /* Initialize SalesForceField fields */
    this._salesForceFieldType = null;
    
    /* Initialize the meta field name TextBox used in form widget designers */
    this._metaFieldNameTextBox = null;

    console.log("Constructor");
    /* Calls the base constructor */
    SalesForce.Fields.DropDownDesigner.initializeBase(this, [element]);
}

SalesForce.Fields.DropDownDesigner.prototype = {
    /* --------------------------------- set up and tear down --------------------------------- */
    initialize: function () {
        /* Here you can attach to events or do other initialization */
        console.log("initialize");
        SalesForce.Fields.DropDownDesigner.callBaseMethod(this, 'initialize');
    },
    dispose: function () {
        /* this is the place to unbind/dispose the event handlers created in the initialize method */
        SalesForce.Fields.DropDownDesigner.callBaseMethod(this, 'dispose');
    },

    /* --------------------------------- public methods ---------------------------------- */

    findElement: function (id) {
        var result = jQuery(this.get_element()).find("#" + id).get(0);
        return result;
    },

    /* Called when the designer window gets opened and here is place to "bind" your designer to the control properties */
    refreshUI: function () {
        var controlData = this._propertyEditor.get_control(); /* JavaScript clone of your control - all the control properties will be properties of the controlData too */

        /* RefreshUI Title */
        jQuery(this.get_title()).val(controlData.Title);

        /* RefreshUI ToolTip */
        jQuery(this.get_note()).val(controlData.Note);

        /* RefreshUI Description */
        jQuery(this.get_description()).val(controlData.Description);

        /* RefreshUI Example */
        jQuery(this.get_example()).val(controlData.Example);



        /* RefreshUI SalesForceFieldType */
        jQuery(this.get_salesForceFieldType()).empty();
        jQuery(this.get_salesForceFieldType()).append('<option value="Contact" selected>Contact</option>');
        jQuery(this.get_salesForceFieldType()).append('<option value="Account">Account</option>');
        jQuery(this.get_salesForceFieldType()).val(controlData.SalesForceFieldType);

        this._loadFields(controlData);

        /* RefreshUI SalesForceField */
        jQuery(this.get_salesForceField()).val(controlData.SalesForceField);

        var that = this;
        jQuery(this.get_salesForceFieldType()).change(function () {
            controlData.SalesForceFieldType = jQuery(that.get_salesForceFieldType()).val();
            that._loadFields(controlData);
        });

    },

    /* Called when the "Save" button is clicked. Here you can transfer the settings from the designer to the control */
    applyChanges: function () {
        var controlData = this._propertyEditor.get_control();

        /* ApplyChanges Title */
        controlData.Title = jQuery(this.get_title()).val();

        /* ApplyChanges ToolTip */
        controlData.Note = jQuery(this.get_note()).val();

        /* ApplyChanges Description */
        controlData.Description = jQuery(this.get_description()).val();

        /* ApplyChanges Example */
        controlData.Example = jQuery(this.get_example()).val();

        /* ApplyChanges SalesForceField */
        controlData.SalesForceField = jQuery(this.get_salesForceField()).val();

        /* ApplyChanges SalesForceAccountsField */
        controlData.SalesForceFieldType = jQuery(this.get_salesForceFieldType()).val();

        controlData.MetaField.FieldName = this.get_metaFieldNameTextBox().get_value();

        controlData.DropDownKey = controlData.SalesForceField;


    },

    /* --------------------------------- event handlers ---------------------------------- */

    /* --------------------------------- private methods --------------------------------- */
    _loadFields: function (controlData) {
        var fields;
        if (controlData.SalesForceFieldType != null &&
             controlData.SalesForceFieldType.toLowerCase() == "account") {
            fields = JSON.parse(controlData.DropDownListAccounts);
        }
        if (controlData.SalesForceFieldType != null &&
            controlData.SalesForceFieldType.toLowerCase() == "contact") {
            fields = JSON.parse(controlData.DropDownListContacts);
        }

        if (fields == "undefined" || fields == null) return;
        jQuery(this.get_salesForceField()).empty();
        var optionElement;
        for (var i = 0; i < fields.length; i++) {
            if (fields[i].name.toLowerCase() == "firstname" ||
                fields[i].name.toLowerCase() == "lastname" ||
                fields[i].name.toLowerCase() == "accountid" ||
                fields[i].name.toLowerCase() == "type__c" ||
                fields[i].name.toLowerCase() == "name" ||
                fields[i].name.toLowerCase() == "physical_country__c") {
                optionElement = jQuery('<option value="' + fields[i].name + '" style="color:red;font-weight: bold;">' + fields[i].name + '</option>');
            } else {
                optionElement = jQuery('<option value="' + fields[i].name + '">' + fields[i].name + '</option>');
            }
            if (fields[i].picklistValues != null && fields[i].picklistValues.length > 0) {
                optionElement.html(optionElement.html() + "<span style='margin-left:10px;color:#e1e1e1;font-weight:bold;'>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;( Drop Down )</span>");
            }

            jQuery(this.get_salesForceField()).append(optionElement);
        }
    },
    /* --------------------------------- properties -------------------------------------- */

    /* Title properties */
    get_title: function () { return this._title; }, 
    set_title: function (value) { this._title = value; },

    /* ToolTip properties */
    get_note: function () { return this._note; }, 
    set_note: function (value) { this._note = value; },

    /* Description properties */
    get_description: function () { return this._description; }, 
    set_description: function (value) { this._description = value; },

    /* Example properties */
    get_example: function () { return this._example; }, 
    set_example: function (value) { this._example = value; },

    /* SalesForceField properties */
    get_salesForceField: function () { return this._salesForceField; }, 
    set_salesForceField: function (value) { this._salesForceField = value; },

    /* SalesForceFieldType properties */
    get_salesForceFieldType: function () { return this._salesForceFieldType; },
    set_salesForceFieldType: function (value) { this._salesForceFieldType = value; },

    /* metaFieldNameTextBox properties */
    get_metaFieldNameTextBox: function () { return this._metaFieldNameTextBox; },
    set_metaFieldNameTextBox: function (value) { this._metaFieldNameTextBox = value; }
}

SalesForce.Fields.DropDownDesigner.registerClass('SalesForce.Fields.DropDownDesigner', Telerik.Sitefinity.Web.UI.ControlDesign.ControlDesignerBase);
